pkg/kine/server/watch.go

```diff
@@ -136,23 +220,85 @@ func (w *watcher) Cancel(watchID int64, err error) {
 	if err != nil {
 		reason = err.Error()
 	}
-	logrus.Debugf("WATCH CANCEL id=%d reason=%s", watchID, reason)
+	logrus.Tracef("WATCH CANCEL id=%d, reason=%s, compactRev=%d", watchID, reason, compactRev)
+
 	serr := w.server.Send(&etcdserverpb.WatchResponse{
-		Header:       &etcdserverpb.ResponseHeader{},
-		Canceled:     true,
-		CancelReason: "watch closed",
-		WatchId:      watchID,
+		Header:          txnHeader(revision),
+		Canceled:        true,
+		CancelReason:    reason,
+		WatchId:         watchID,
+		CompactRevision: compactRev,
 	})
-	if serr != nil && err != nil {
+	if serr != nil && err != nil && !clientv3.IsConnCanceled(serr) {
 		logrus.Errorf("WATCH Failed to send cancel response for watchID %d: %v", watchID, serr)
 	}
 }
 
 func (w *watcher) Close() {
+	logrus.Tracef("WATCH SERVER CLOSE")
 	w.Lock()
-	for _, v := range w.watches {
-		v()
+	for id, progressCh := range w.progress {
+		close(progressCh)
+		delete(w.progress, id)
+	}
+	for id, cancel := range w.watches {
+		cancel()
+		delete(w.watches, id)
 	}
 	w.Unlock()
 	w.wg.Wait()
 }
+
+// Progress sends a progress report if all watchers are synced.
+// Ref: https://github.com/etcd-io/etcd/blob/v3.5.11/server/mvcc/watchable_store.go#L500-L504
+func (w *watcher) Progress(ctx context.Context) {
+	w.RLock()
```

Comment:
While it might be sound in this case, taking an RLock and than writing to a field (posting to the progress channel) sounds like a bug IMHO. If we are reading, we should really not write.